File: internal/tools/slice.go
package tools

import (
	"context"
	"strings"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/greentech/gt-data-1c/internal/refusal"
)

// ── Срез регистра сведений ────────────────────────────────────────────────────

type SliceInput struct {
	Base       string         `json:"base,omitempty" jsonschema:"Имя базы 1С из реестра. Опущено — база по умолчанию"`
	Register   string         `json:"register" jsonschema:"Имя регистра сведений без префикса, например ЦеныНоменклатуры"`
	Kind       string         `json:"kind,omitempty" jsonschema:"СрезПоследних (умолчание) или СрезПервых"`
	Period     string         `json:"period,omitempty" jsonschema:"Дата среза, например 2026-06-30. Пусто — на текущий момент"`
	Where      string         `json:"where,omitempty" jsonschema:"Отбор по полям среза, без слова ГДЕ. Пример: Номенклатура = &Товар"`
	Parameters map[string]any `json:"parameters,omitempty" jsonschema:"Параметры отбора: ключ без амперсанда"`
	Limit      int            `json:"limit,omitempty" jsonschema:"Сколько строк вернуть (по умолчанию 100, максимум 1000)"`
}

func SliceTool() *mcp.Tool {
	return &mcp.Tool{
		Name: "slice",
		Description: "Получить срез регистра сведений — значения, действующие на дату: цены " +
			"номенклатуры, курсы валют, ставки, настройки. Запрос собирается сервером: измерения " +
			"и ресурсы регистра подставляются сами, дата уходит параметром среза. Используй это " +
			"вместо запроса к самому регистру: выборка записей вернёт всю историю, и первая " +
			"попавшаяся строка легко сойдёт за действующее значение. Непериодический регистр " +
			"даёт отказ — у него среза не бывает.",
	}
}

func (s *Set) Slice(ctx context.Context, _ *mcp.CallToolRequest, in SliceInput) (*mcp.CallToolResult, any, error) {
	if strings.TrimSpace(in.Register) == "" {
		return nil, nil, refusal.New(refusal.BadRequest, "регистр не назван", "поле register обязательно")
	}
	client, err := s.channelFor(in.Base)
	if err != nil {
		return nil, nil, err
	}

	payload := map[string]any{"register": in.Register}
	for key, value := range map[string]string{
		"kind": in.Kind, "period": in.Period, "where": in.Where,
	} {
		if strings.TrimSpace(value) != "" {
			payload[key] = value
		}
	}
	if len(in.Parameters) > 0 {
		payload["parameters"] = in.Parameters
	}
	if in.Limit > 0 {
		payload["limit"] = in.Limit
	}

	var reply queryReply
	if err := client.Tell(ctx, "slice", payload, &reply); err != nil {
		return nil, nil, err
	}
	return text(renderTable(client.Base().Name, reply)), nil, nil
}

// ── Итоги по счетам ───────────────────────────────────────────────────────────

type AccountsInput struct {
	Base       string         `json:"base,omitempty" jsonschema:"Имя базы 1С из реестра. Опущено — база по умолчанию"`
	Account    string         `json:"account" jsonschema:"Код счёта, например 41 или 62.01"`
	Kind       string         `json:"kind,omitempty" jsonschema:"Остатки (умолчание), Обороты или ОстаткиИОбороты"`
	Period     string         `json:"period,omitempty" jsonschema:"Дата остатков для kind=Остатки. Пусто — на текущий момент"`
	Start      string         `json:"start,omitempty" jsonschema:"Начало периода для Обороты и ОстаткиИОбороты"`
	End        string         `json:"end,omitempty" jsonschema:"Конец периода для Обороты и ОстаткиИОбороты"`
	Register   string         `json:"register,omitempty" jsonschema:"Имя регистра бухгалтерии (по умолчанию Хозрасчетный)"`
	Resources  []string       `json:"resources,omitempty" jsonschema:"Показатели: СуммаОстатокДт, СуммаОборотДт и т.п. Пусто — стандартный набор для выбранного вида"`
	Parameters map[string]any `json:"parameters,omitempty" jsonschema:"Параметры отбора: ключ без амперсанда"`
	Limit      int            `json:"limit,omitempty" jsonschema:"Сколько строк вернуть (по умолчанию 100, максимум 1000)"`
}

func AccountsTool() *mcp.Tool {
	return &mcp.Tool{
		Name: "accounts",
		Description: "Получить бухгалтерские итоги по счёту: остатки на дату (kind=Остатки), обороты " +
			"за период (kind=Обороты) или полную картину с начальным и конечным остатком " +
			"(kind=ОстаткиИОбороты). Счёт задаётся кодом — 41, 62.01, 51. Суммы приходят раздельно " +
			"по дебету и кредиту. Это для конфигураций с бухгалтерским учётом; складские и товарные " +
			"итоги живут в регистрах накопления, их берёт инструмент register.",
	}
}

func (s *Set) Accounts(ctx context.Context, _ *mcp.CallToolRequest, in AccountsInput) (*mcp.CallToolResult, any, error) {
	if strings.TrimSpace(in.Account) == "" {
		return nil, nil, refusal.New(refusal.BadRequest, "счёт не назван",
			"поле account обязательно: код счёта, например 41 или 62.01",
			"перечень счетов — metadata с filter=ПланыСчетов, затем object по нужному плану")
	}
	client, err := s.channelFor(in.Base)
	if err != nil {
		return nil, nil, err
	}

	payload := map[string]any{"account": in.Account}
	for key, value := range map[string]string{
		"kind": in.Kind, "period": in.Period, "start": in.Start,
		"end": in.End, "register": in.Register,
	} {
		if strings.TrimSpace(value) != "" {
			payload[key] = value
		}
	}
	if len(in.Resources) > 0 {
		payload["resources"] = in.Resources
	}
	if len(in.Parameters) > 0 {
		payload["parameters"] = in.Parameters
	}
	if in.Limit > 0 {
		payload["limit"] = in.Limit
	}

	var reply queryReply
	if err := client.Tell(ctx, "accounts", payload, &reply); err != nil {
		return nil, nil, err
	}
	return text(renderTable(client.Base().Name, reply)), nil, nil
}

File: crates/gf-data-1c/src/tools/slice.rs
//! Срез регистра сведений и бухгалтерские итоги по счёту.
//!
//! Оба инструмента собирают текст запроса на стороне сервера. Смысл тот же, что у
//! `register`: имя виртуальной таблицы, порядок границ периода и подстановка измерений
//! не сочиняются заново — их нельзя ошибиться.

use std::collections::BTreeMap;

use serde::Deserialize;
use serde_json::{json, Map, Value};

use crate::refusal::{Kind, Refusal};

use super::data::QueryReply;
use super::lenient::StringList;
use super::Set;

// ── Срез регистра сведений ───────────────────────────────────────────────────

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct SliceInput {
    pub base: String,
    /// Имя регистра сведений без префикса, например ЦеныНоменклатуры.
    pub register: String,
    /// `СрезПоследних` (умолчание) или `СрезПервых`.
    pub kind: String,
    /// Дата среза. Пусто — на текущий момент.
    pub period: String,
    /// Отбор по полям среза, без слова ГДЕ.
    pub where_: String,
    pub parameters: BTreeMap<String, Value>,
    pub limit: i64,
}

pub const SLICE_NAME: &str = "slice";
pub const SLICE_DESCRIPTION: &str = "Получить срез регистра сведений — значения, действующие \
на дату: цены номенклатуры, курсы валют, ставки, настройки. Запрос собирается сервером: \
измерения и ресурсы регистра подставляются сами, дата уходит параметром среза. Используй это \
вместо запроса к самому регистру: выборка записей вернёт всю историю, и первая попавшаяся \
строка легко сойдёт за действующее значение. Непериодический регистр даёт отказ — у него \
среза не бывает.";

// ── Итоги по счетам ──────────────────────────────────────────────────────────

#[derive(Debug, Default, Deserialize)]
#[serde(default)]
pub struct AccountsInput {
    pub base: String,
    /// Код счёта, например 41 или 62.01.
    pub account: String,
    /// `Остатки` (умолчание), `Обороты` или `ОстаткиИОбороты`.
    pub kind: String,
    pub period: String,
    pub start: String,
    pub end: String,
    /// Разбивка периода: Месяц, Квартал, Год, День, Регистратор.
    pub periodicity: String,
    /// Имя регистра бухгалтерии (по умолчанию Хозрасчетный).
    pub register: String,
    /// Показатели: СуммаОстатокДт, СуммаОборотДт и т.п.
    pub resources: StringList,
    pub parameters: BTreeMap<String, Value>,
    pub limit: i64,
}

pub const ACCOUNTS_NAME: &str = "accounts";
pub const ACCOUNTS_DESCRIPTION: &str = "Получить бухгалтерские итоги по счёту: остатки на дату \
(kind=Остатки), обороты за период (kind=Обороты) или полную картину с начальным и конечным \
остатком (kind=ОстаткиИОбороты). Счёт задаётся кодом — 41, 62.01, 51. Разбивку по периодам \
даёт periodicity (Месяц, Квартал, Год) — так собирается движение счёта помесячно, без единой \
строки на языке запросов. Суммы приходят раздельно по дебету и кредиту. Это для конфигураций \
с бухгалтерским учётом; складские и товарные итоги живут в регистрах накопления, их берёт \
инструмент register.";

impl Set {
    /// Срез регистра сведений на дату.
    pub fn slice(&self, input: &SliceInput) -> Result<String, Refusal> {
        if input.register.trim().is_empty() {
            return Err(Refusal::new(
                Kind::BadRequest,
                "регистр не назван",
                "поле register обязательно",
            ));
        }
        let client = self.channel_for(&input.base)?;

        let mut payload = Map::new();
        payload.insert("register".into(), json!(input.register));
        for (key, value) in [
            ("kind", &input.kind),
            ("period", &input.period),
            ("where", &input.where_),
        ] {
            if !value.trim().is_empty() {
                payload.insert(key.into(), json!(value));
            }
        }
        if !input.parameters.is_empty() {
            payload.insert("parameters".into(), json!(input.parameters));
        }
        if input.limit > 0 {
            payload.insert("limit".into(), json!(input.limit));
        }

        let reply: QueryReply = client.tell("slice", &Value::Object(payload))?;
        Ok(super::data::render_table(&client.base().name, &reply))
    }

    /// Бухгалтерские итоги по счёту.
    pub fn accounts(&self, input: &AccountsInput) -> Result<String, Refusal> {
        if input.account.trim().is_empty() {
            return Err(Refusal::new(
                Kind::BadRequest,
                "счёт не назван",
                "поле account обязательно: код счёта, например 41 или 62.01",
            )
            .hint(
                "перечень счетов — metadata с filter=ПланыСчетов, затем object по нужному плану",
            ));
        }
        let client = self.channel_for(&input.base)?;

        let mut payload = Map::new();
        payload.insert("account".into(), json!(input.account));
        for (key, value) in [
            ("kind", &input.kind),
            ("period", &input.period),
            ("start", &input.start),
            ("end", &input.end),
            ("register", &input.register),
            ("periodicity", &input.periodicity),
        ] {
            if !value.trim().is_empty() {
                payload.insert(key.into(), json!(value));
            }
        }
        if !input.resources.is_empty() {
            payload.insert("resources".into(), json!(input.resources.as_slice()));
        }
        if !input.parameters.is_empty() {
            payload.insert("parameters".into(), json!(input.parameters));
        }
        if input.limit > 0 {
            payload.insert("limit".into(), json!(input.limit));
        }

        let reply: QueryReply = client.tell("accounts", &Value::Object(payload))?;
        Ok(super::data::render_table(&client.base().name, &reply))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_set(tag: &str) -> Set {
        let dir = std::env::temp_dir().join(format!("gfdata-slice-{}-{tag}", std::process::id()));
        let _ = std::fs::create_dir_all(&dir);
        let path = dir.join("bases.json");
        let _ = std::fs::remove_file(&path);
        let mut s = Set::new("0.1.0");
        s.registry_path = Some(path);
        s
    }

    #[test]
    fn регистр_обязателен_для_среза() {
        let err = make_set("срез").slice(&SliceInput::default()).unwrap_err();
        assert_eq!(err.kind, Kind::BadRequest);
        assert!(err.to_string().contains("регистр не назван"), "{err}");
    }

    #[test]
    fn счёт_обязателен_и_отказ_говорит_где_взять_перечень() {
        let err = make_set("счёт")
            .accounts(&AccountsInput::default())
            .unwrap_err();
        assert_eq!(err.kind, Kind::BadRequest);
        assert!(
            err.to_string().contains("62.01"),
            "пример важнее правила: {err}"
        );
        assert!(err.to_string().contains("metadata"), "{err}");
    }
}
